woocommerce/woocommerce
# Update Checkout Fields error reporting to use `_doing_it_wrong` instead of wc logger

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Update Checkout Fields error reporting to use `_doing_it_wrong` instead of wc logger.

Previously we were using the WC logger which caused repetitive error logging as the fields are registered on every page load.

Closes #43277



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Add this snippet to your site somewhere and be logged in as admin

```php
add_action(
	'woocommerce_loaded',
	function() {

		woocommerce_blocks_register_checkout_field(
			array(
				'id'             => 'plugin-namespace/test-field',
				'label'          => __( 'Test field', 'woo-gutenberg-products-block' ),
				'optionalLabel'  => __( 'Test field', 'woo-gutenberg-products-block' ),
				'autocomplete'   => 'test-field',
				'autocapitalize' => 'characters',
				'location'       => 'address',
				'type'           => 'checkbox',
			)
		);
	}
);
```

1. Load the checkout block and ensure no errors appear.
2. In the snippet, remove the `id` key.
3. Load the checkout block and expect to see a "doing it wrong" error.
4. Replace the `id` key and remove the `label` key.
5. Load the checkout block and expect to see a "doing it wrong" error.
7. Replace the `label` key and change the value to an empty string and then null. Both cases should yield a doing it wrong error.
8. Set it back to a string.
9. Remove the `location` key. 
10. Load the checkout block and expect to see a "doing it wrong" error.
11. Replace the `location` key and change the value to `test`
12. Load the checkout block and expect to see a "doing it wrong" error.
13. Reset the `location` key's value to `address`
14. Remove the `type` key
15. Load the checkout block and expect to see a "doing it wrong" error.
16. Replace the `type` key and set its value to `test`
17. Load the checkout block and expect to see a "doing it wrong" error.
18. Reset the `type` key's value to `checkbox`.
19. Add a new key: `'required' => true`
20. Load the checkout block and expect to see a "doing it wrong" error. (For now, we won't allow mandatory checkboxes, this will be done in a follow up. This test is just to check for correct error reporting!)
21. Change the type to `select` and set `required` to false.
22. Add the following key for the select options:
```php
				'options'  => [
					[
						'label' => 'Yes',
						'value' => 'yes',
					],
					[
						'label' => 'No',
						'value' => 'no',
					],
				]
```

24. Load the checkout block and expect to see a "doing it wrong" error. (For now, we won't allow optional selects, this will be done in a follow up. This test is just to check for correct error reporting!)
25. Duplicate the snippet so that you're registering the same field twice.
26. Load the checkout block and expect to see a "doing it wrong" error. 
27. Remove one of the duplicates.
28.  In the `options` key, add a new key with the same value as another one, for example:
```php

				'options'  => [
					[
						'label' => 'Yes',
						'value' => 'yes',
					],
					[
						'label' => 'No',
						'value' => 'no',
					],
					[
						'label' => 'Also no',
						'value' => 'no',
					],
				]
```
26. Load the checkout block and expect to see a "doing it wrong" error. 
27. Remove the `options` key completely.
28. Load the checkout block and expect to see a "doing it wrong" error. 
29. Finally, change the `type` to `hidden`
30. Load the checkout block and expect to see a "doing it wrong" error. 




### Changelog entry




<details>

#### Significance




#### Type




#### Message 

Change the way errors are reported when incorrectly registering custom checkout fields.

#### Comment 

</details>
